File: app/src/main/kotlin/tech/mmarca/openvitals/data/repository/contract/HeartRepository.kt
package tech.mmarca.openvitals.data.repository.contract

import java.time.Instant
import java.time.LocalDate
import tech.mmarca.openvitals.core.period.PeriodLoadQuery
import tech.mmarca.openvitals.data.repository.HeartPeriodMetric
import tech.mmarca.openvitals.domain.model.DailyHrv
import tech.mmarca.openvitals.domain.model.DailyRestingHR
import tech.mmarca.openvitals.domain.model.HeartRateSample
import tech.mmarca.openvitals.domain.model.HeartRateSummary
import tech.mmarca.openvitals.domain.model.HrvSample
import tech.mmarca.openvitals.domain.query.HeartPeriodData

interface HeartRepository {
    suspend fun loadHeartPeriod(
        query: PeriodLoadQuery,
        metric: HeartPeriodMetric,
    ): HeartPeriodData

    /** The day's average as the Today tile shows it: raw samples, minute-bucketed. */
    suspend fun loadAvgHeartRate(date: LocalDate): Long?

    suspend fun loadRawHeartRateSamplesForDayGraph(date: LocalDate): List<HeartRateSample>

    suspend fun loadHeartRateSamples(start: LocalDate, end: LocalDate): List<HeartRateSample>

    suspend fun loadHeartRateSamples(start: Instant, end: Instant): List<HeartRateSample>

    /** The highest bpm in the window, without loading its samples. Any window length is safe. */
    suspend fun loadMaxHeartRate(start: Instant, end: Instant): Long?

    suspend fun loadDailyHeartRateSummaries(start: LocalDate, end: LocalDate): List<HeartRateSummary>

    suspend fun loadRestingHeartRate(date: LocalDate): Long?

    suspend fun loadDailyRestingHR(start: LocalDate, end: LocalDate): List<DailyRestingHR>

    /** The day's HRV as the Day view shows it: raw samples, minute-bucketed. */
    suspend fun loadHrvRmssd(date: LocalDate): Double?

    suspend fun loadHrvSamples(start: Instant, end: Instant): List<HrvSample>

    suspend fun loadDailyHRV(start: LocalDate, end: LocalDate): List<DailyHrv>
}

File: lib/data/repository/contract/heart_repository.dart
import '../../../core/period/period_load_query.dart';
import '../../../core/result/result.dart';
import '../../../core/time/local_date.dart';
import '../../../domain/model/heart_models.dart';
import '../../../domain/model/refresh_mode.dart';
import '../../../domain/query/heart_period_data.dart';

/// Which heart metric family a period load should populate. Port of the Kotlin
/// `HeartPeriodMetric` (declared alongside `HeartRepositoryImpl`).
enum HeartPeriodMetric { all, averageHeartRate, restingHeartRate, hrv }

/// Port of the Kotlin `HeartRepository` contract.
///
/// Fallible operations return [Result]; every method here is a Health Connect
/// read, so there is no synchronous cached-state probe to stay bare.
abstract interface class HeartRepository {
  Future<Result<HeartPeriodData>> loadHeartPeriod(
    PeriodLoadQuery query,
    HeartPeriodMetric metric, {
    RefreshMode refreshMode = RefreshMode.normal,
  });

  Future<Result<List<HeartRateSample>>> loadHeartRateSamplesForDay(
    LocalDate date,
  );

  Future<Result<List<HeartRateSample>>> loadRawHeartRateSamplesForDayGraph(
    LocalDate date,
  );

  Future<Result<List<HeartRateSample>>> loadHeartRateSamples(
    LocalDate start,
    LocalDate end,
  );

  Future<Result<List<HeartRateSample>>> loadHeartRateSamplesInstant(
    DateTime start,
    DateTime end,
  );

  Future<Result<List<HeartRateSummary>>> loadDailyHeartRateSummaries(
    LocalDate start,
    LocalDate end,
  );

  Future<Result<int?>> loadRestingHeartRate(LocalDate date);

  Future<Result<List<DailyRestingHR>>> loadDailyRestingHR(
    LocalDate start,
    LocalDate end,
  );

  Future<Result<double?>> loadHrvRmssd(LocalDate date);

  Future<Result<List<HrvSample>>> loadHrvSamples(DateTime start, DateTime end);

  Future<Result<List<DailyHrv>>> loadDailyHRV(LocalDate start, LocalDate end);
}
